sdk/azidentity/authorization_code_credential.go

```diff
@@ -44,32 +47,70 @@ func NewAuthorizationCodeCredential(tenantID string, clientID string, authCode s
 	if !validTenantID(tenantID) {
 		return nil, errors.New(tenantIDValidationErr)
 	}
-	cp := AuthorizationCodeCredentialOptions{}
-	if options != nil {
-		cp = *options
+	if options == nil {
+		options = &AuthorizationCodeCredentialOptions{}
 	}
-	authorityHost, err := setAuthorityHost(cp.AuthorityHost)
+	authorityHost, err := setAuthorityHost(options.AuthorityHost)
 	if err != nil {
 		return nil, err
 	}
-	c, err := newAADIdentityClient(authorityHost, &cp.ClientOptions)
+	if options.ClientSecret != "" {
+		cred, err := confidential.NewCredFromSecret(options.ClientSecret)
+		if err != nil {
+			return nil, err
+		}
+		c, err := confidential.New(clientID, cred,
+			confidential.WithAuthority(runtime.JoinPaths(authorityHost, tenantID)),
+			confidential.WithHTTPClient(newPipelineAdapter(&options.ClientOptions)),
+		)
+		if err != nil {
+			return nil, err
+		}
+		return &AuthorizationCodeCredential{authCode: authCode, clientSecret: options.ClientSecret, redirectURI: redirectURL, cca: c}, nil
+	}
+	c, err := public.New(clientID,
+		public.WithAuthority(runtime.JoinPaths(authorityHost, tenantID)),
+		public.WithHTTPClient(newPipelineAdapter(&options.ClientOptions)),
+	)
 	if err != nil {
 		return nil, err
 	}
-	return &AuthorizationCodeCredential{tenantID: tenantID, clientID: clientID, authCode: authCode, clientSecret: cp.ClientSecret, redirectURI: redirectURL, client: c}, nil
+	return &AuthorizationCodeCredential{authCode: authCode, clientSecret: options.ClientSecret, redirectURI: redirectURL, pca: c}, nil
 }
 
 // GetToken obtains a token from Azure Active Directory by redeeming the authorization code. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *AuthorizationCodeCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	tk, err := c.client.authenticateAuthCode(ctx, c.tenantID, c.clientID, c.authCode, c.clientSecret, "", c.redirectURI, opts.Scopes)
+	if c.cca != nil {
+		ar, err := c.cca.AcquireTokenSilent(ctx, opts.Scopes, confidential.WithSilentAccount(c.account))
```

Comment:
I bet you could place this common code in a local func.
```go
acquireToken := func(func acquireSilent(/*params*/), func acquireByAuthCode(/*params*/)) (msal.AuthenticationResult, error) {
// stuff....
}
var ar msal.AuthenticationResult
var err error
if c.cca != nil {
    ar, err = acquireToken(c.cca.AcquireTokenSilent, c.cca.AcquireTokenByAuthCode)
}
// etc...
```

Reply:
The sticky part is that the MSAL methods have incompatible options params, e.g. `public.AcquireTokenSilent` requires `public.AcquireTokenSilentOption`, so a function which might call `confidential.AcquireTokenSilent` instead must know the difference. I guess I could hide the difference behind another type but imagine that isn't worth the cost in legibility. What do you think of something like this:
```go
var acquireTokenSilent func() (base.AuthResult, error)
if c.cca != nil {
    acquireTokenSilent = func() (base.AuthResult, error) {
        return c.cca.AcquireTokenSilent(ctx, opts.Scopes, confidential.WithSilentAccount(c.account))
    }
} else {
    acquireTokenSilent = func() (base.AuthResult, error) {
        return c.pca.AcquireTokenSilent(ctx, opts.Scopes, public.WithSilentAccount(c.account))
    }
}
```
Still repetitive but less so, and allows sharing the business logic at least.

Reply:
Ah good point, I didn't look super close.  What you have is fine.